Reject invalid subscription urls

allenpetersen opened the pull request:
###### Required for all non-trivial PRs
- [x] Rebased/mergable
- [x] Tests pass
- [x] CHANGELOG.md updated
- [x] Sign [CLA](https://influxdata.com/community/cla/) (if not already signed)

The url must have a scheme of udp or tcp in addition to a port number.
CREATE SUBSCRIPTION will fail if there are invalid destinations.

Additionally Service.createSubscription will skip over any invalid
destinations on previously created subscriptions.

Fixes #7615

allenpetersen commented:
Is a useful default protocol and port number to use for subscriptions? The current docs and tests fully qualify but use udp and port 9090. Also the docs describe the destinations as "host" so I could see a user just putting in a hostname and expecting it to work.

allenpetersen commented:
The InfluxDB glossary says "subscriptions can use TCP or UDP for transmitting the writes." Service.newPointsWriter() only accepts udp/http/https. I wrote validateURL based on the glossary docs. Should the docs and validate function change to match newPointsWriter()?

e-dard commented on CHANGELOG.md:
```diff
@@ -13,6 +13,7 @@
 - [#7621](https://github.com/influxdata/influxdb/issues/7621): Expand string and boolean fields when using a wildcard with `sample()`.
 - [#7616](https://github.com/influxdata/influxdb/pull/7616): Fix chuid argument order in init script @ccasey
 - [#7656](https://github.com/influxdata/influxdb/issues/7656): Fix cross-platform backup/restore
+- [#7615](https://github.com/influxdata/influxdb/issues/7615): Reject invalid subscription urls.
```
Please feel free to give yourself credit (`add @allenpetersen`)

e-dard commented on services/meta/data.go:
```diff
@@ -465,8 +466,33 @@ func (data *Data) DropContinuousQuery(database, name string) error {
+func validateURL(input string) error {
+	u, err := url.Parse(input)
+	if err != nil {
+		fmt.Println(err)
```
Remove this.

allenpetersen replied:
oops, forgot to remove this debug logging. fixed.

e-dard commented on services/meta/data.go:
```diff
@@ -465,8 +466,33 @@ func (data *Data) DropContinuousQuery(database, name string) error {
+		return ErrInvalidSubscriptionURL(input)
+	}
+
+	hostParts := strings.Split(u.Host, ":")
```
I'm not sure what the intention is here. Are you checking that a port has been set on the URL?

allenpetersen replied:
The issue addresses having an invalid scheme/protocol but using udp without knowing the port number is also not correct.

Currently it just requires a non-empty string to be set for the port. This is related to my questions about default port numbers and tcp vs http. Once I better understand the requirements this will most likely change, at a minimum it could verify the port is a positive integer.

e-dard replied:
[net.SplitHostPort](https://golang.org/pkg/net/#SplitHostPort) might be a better fit, here.

e-dard commented on services/meta/data.go:
```diff
@@ -465,8 +466,33 @@ func (data *Data) DropContinuousQuery(database, name string) error {
 func (data *Data) CreateSubscription(database, rp, name, mode string, destinations []string) error {
+	for _, d := range destinations {
+		if err := validateURL(d); err != nil {
+			return ErrInvalidSubscriptionURL(d)
```
`validateURL` only returns an error created with `ErrInvalidSubscriptionURL` so I think you can just `return err` here.

e-dard commented on services/subscriber/service.go:
```diff
@@ -183,20 +183,28 @@ func (s *Service) createSubscription(se subEntry, mode string, destinations []st
-	writers := make([]PointsWriter, len(destinations))
-	stats := make([]writerStats, len(writers))
-	for i, dest := range destinations {
+	writers := []PointsWriter{}
```
Why have you changed these? Usually slices have been pre-allocated for a reason.

e-dard replied:
You can pre-allocate and still use `append`. You just need to allocate the capacity and a length of zero. For example: `writers := make([]PointsWriter, 0, len(destinations))`

allenpetersen replied:
I think using a slice literal is cleaner and easier to read. It is my default unless there is a clear performance reason to do otherwise.

Checking now I notice that using make with a zero length is a common pattern in the InfluxDB codebase. I'll change these lines.

e-dard commented on services/subscriber/service.go:
```diff
@@ -183,20 +183,28 @@ func (s *Service) createSubscription(se subEntry, mode string, destinations []st
 		if err != nil {
-			return nil, err
+			s.Logger.Println("failed to parse subscription url:", dest)
+			continue
```
What's the rationale for ignoring (aside from logging the error) any invalid subscriptions? It's unlikely operators are going to notice that their subscriptions have not been added unless they're checking the logs.

e-dard commented on services/subscriber/service.go:
```diff
@@ -183,20 +183,28 @@ func (s *Service) createSubscription(se subEntry, mode string, destinations []st
 		u, err := url.Parse(dest)
 		if err != nil {
-			return nil, err
+			s.Logger.Println("failed to parse subscription url:", dest)
```
Log messages should be correct sentences, and `Printf` is more common in the project, i.e.,

```go
s.Logger.Printf("Failed to parse subscription url: %s", dest)
```

Same issue applies to all other log messages in the PR.

allenpetersen replied:
I first wrote it similar to your suggestion but noticed it didn't match the rest of the file. Being new to the project I just followed the pattern. There are a 11 Logger calls in the file, all of them use Println and most are sentence fragments. I've fixed my log messages.

e-dard commented on services/subscriber/service.go:
```diff
@@ -285,7 +293,8 @@ func (s *Service) updateSubs(wg *sync.WaitGroup) error {
 				sub, err := s.createSubscription(se, si.Mode, si.Destinations)
 				if err != nil {
-					return err
+					s.Logger.Println("createSubscription failed for:", si.Name)
```
`s.Logger.Printf("Subscription creation failed for: %s", si.Name)`

e-dard replied:
Again, shouldn't this be an error? Why would it be sensible to continue after encountering this?

allenpetersen replied:
Returning the error will only silently stop processing the rest of the subscriptions. The new code will log the error and continue setting up the rest of the subscriptions.

e-dard commented:
@allenpetersen we don't support raw `tcp` so the implication in the docs is that subscriptions can use UDP or HTTP/HTTPS (TCP).

e-dard commented:
@nathanielc can you take a look at this please?

allenpetersen commented:
I removed tcp and added http(s). I suppose that if they are using http(s) there are known default ports but Kapacitor's default port is 9092 so using 80/443 is most likely not what the user wants to do.

Should create subscription require a fully qualified `protocol://host:port` destination or apply some defaults for parts that are not supplied?

nathanielc commented:
>Should create subscription require a fully qualified protocol://host:port destination or apply some defaults for parts that are not supplied?

For now I would require a fully qualified `protocol://host:port`. If we find a useful set of defaults they can be added later.

nathanielc commented on services/subscriber/service.go:
```diff
@@ -285,7 +293,8 @@ func (s *Service) updateSubs(wg *sync.WaitGroup) error {
 				if err != nil {
-					return err
+					s.Logger.Printf("Subscription creation failed for: %s", si.Name)
+					continue
```
Now that no error is returned we should change the return of the updateSubs method to return nothing, since there is no longer any case where an error is returned.

f we run into this error it means that we have a configured subscription but it is not running. Since there is no way to query the status of the subscriptions we should add a statistic for this error so that it is easier to detect. That way a user can setup an alert on the statistic to know if/when subscriptions are failing.

allenpetersen replied:
I removed the unnecessary error return value and a new `createFailures` statistic.

nathanielc replied:
I expected the `CreateFailures` stat to be incremented after a call to `createSubscription` so that one call can increment the stat by at most one. Can you change it so that the stat is incremented only after the call to `createSubscriptions`. 

If we also want to track the failures per destination then a new stats should be added to the `writerStats` type.

allenpetersen replied:
I simplified things a little since the new validation should prevent us from getting into this code path. Subscriptions are back to being 'all or nothing' but one bad subscription doesn't prevent others from being created.

If a subscription has a single bad destination it fails with an error and the `CreateFailures` stat will get incremented. Now there is no need to track the per destination failures.

e-dard commented on services/subscriber/service.go:
```diff
@@ -183,20 +186,22 @@ func (s *Service) createSubscription(se subEntry, mode string, destinations []st
 		u, err := url.Parse(dest)
 		if err != nil {
-			return nil, err
+			return nil, fmt.Errorf("Failed to parse subscription url: %s", dest)
```
Error messages should not be capitalised. `s/Failed/failed` Same thing a few lines down

allenpetersen commented:
@e-dard @nathanielc I think I've addressed all of the review comments. There was a conflict with the changelog so I rebased the commits.

e-dard commented:
@nathanielc can you approve if this now LGTY.

nathanielc commented:
@allenpetersen @e-dard Looks like the error messages are still capitalized, maybe the rebase messed up the intended change. Other than that LGTM.

allenpetersen commented:
@nathanielc Yes, I did miss a commit on my last rebase. I've rebased the branch again and it should be good to go now.

nathanielc commented:
@e-dard LGTM

e-dard commented:
Nice work @allenpetersen; thanks!
